### PR TITLE
Raise warning if proxy key is eg. "all" instead of "all://".

### LINKED ISSUE #1105: On `proxies={"http": ...}` vs. `proxies={"http://": ...}`
Currently we're in line with `requests` in supporting "all", "http", and "https" as valid proxy keys.

Internally we handle those as URL patterns that we can match against, and eg. `"http"` is *actually* mapped to `"http://"`.

I'm wondering if we might have better internal consistency by...

* Raising a warning on base usage of "all", "http", "https", later escalating this to an error.
* Instead requiring an explicit "all://", "http://", "https://", all of which properly parse directly as URLs.

My feeling is that doing so would create a nicer consistency, and also has a subtle indication to developers about what is actually going on here.

For example...

`proxies={"http://": httpx.Proxies(...)}` neatly implies that `proxies={"http://www.example.com": httpx.Proxies(...)}` might also be a valid thing to do, which indeed it is.

Allowing `proxies={"http": httpx.Proxies(...)}` makes it less obvious that the key is a general purpose URL pattern.

~~Additionally this change means that other usages will end up with more expected behaviors, eg. `proxies={"www.example.com": ...}` will do what you'd expect, rather than mapping to `proxies={"www.example.com://": ...}` which is what we currently do.~~ - Not true, that'd parse as scheme='', host='', path='www.example.com'. Instead we'd probably want to be strict and ensure that only absolute URLs are allowed as pattern keys. (While still including a graceful deprecation for plain "http", "https", "all")

This becomes doubly relevant once we introduce a mount API, which will use the same URL pattern style for it's API, as current `proxies=...` argument.

### PR DESCRIPTION
**Pull request by cdeler**
Closes #1105

### CODE REVIEW

**Comment by cdeler** [comment]
I see the tests, should be changed, also I can add tests for the new warning appearing. @tomchristie let me know if I can continue with that

**Comment by tomchristie** [comment]
Thanks for taking a go at this!

So, I'd change the implementation here, for simplicity.

Instead of changing *anything* in the Client, I'd add a single line here... https://github.com/encode/httpx/blob/78cf16ace92b42de5972a490787002631d558b32/httpx/_utils.py#L481 
that raises a warning such as `f"Proxy keys should use proper URL forms rather than plain scheme strings. Instead of {pattern}, use {pattern}://"` - that ought to be enough to deal with this more neatly.

You probably also want to then update the docs in https://github.com/encode/httpx/blob/master/docs/advanced.md to use the preferred styles.

File: tests/client/test_proxies.py [hunk 1]
```diff
@@ -244,3 +247,21 @@ def test_proxies_environ(monkeypatch, client_class, url, env, expected):
...
+    ],
+)
+def test_for_deprecated_proxy_params(proxies, expected_scheme):
+    with pytest.deprecated_call() as block:
+        httpx.AsyncClient(proxies=proxies)
+
+    warning_message = str(block.pop(DeprecationWarning))
+
+    for scheme in expected_scheme:
+        assert scheme in warning_message
```
Comment (florimondmanca) [inline tests/client/test_proxies.py:267]: Could this be simplified to this by converting `expected_scheme` to a plain string?

```suggestion
    assert expected_scheme in warning_message
```

Comment (tomchristie) [reply tests/client/test_proxies.py:267]: Yeah I think that was walking through the *characters* in the string. :)

File: tests/client/test_proxies.py [hunk 2]
```diff
@@ -244,3 +247,20 @@ def test_proxies_environ(monkeypatch, client_class, url, env, expected):
...
     else:
         assert transport.proxy_origin == url_to_origin(expected)
+
+
+@pytest.mark.parametrize(
+    ["proxies", "expected_scheme"],
+    [
+        ({"http": "http://127.0.0.1"}, ["http://"]),
+        ({"https": "http://127.0.0.1"}, ["https://"]),
+        ({"all": "http://127.0.0.1"}, ["all://"]),
```
Comment (florimondmanca) [inline tests/client/test_proxies.py:257]:
```suggestion
        ({"http": "http://127.0.0.1"}, "http://"),
        ({"https": "http://127.0.0.1"}, "https://"),
        ({"all": "http://127.0.0.1"}, "all://"),
```

File: tests/client/test_proxies.py (hunk 1 above)
Comment (florimondmanca) [reply tests/client/test_proxies.py]: No, it was correct (I expect CI to fail now! ;-)), but `expected_scheme` was unecessarily a 1-item list. Sent a patch…

**Pull request merged**